### PR TITLE
Implement estimateOverlappingEntries in fenced files

### LINKED ISSUE #4611: Estimation of rfile entries is not implemented for fenced files.
**Describe the bug**

When merging #4452 from 2.1 to main in 72110813107f14949bc082a109b4e35b06ea39ad estimating entries for a fenced rfile was not implemented.   Wanted to avoid making this change in a merge commit.  

**Expected behavior**

[Rfile.FencedReader.estimateOverlappingEntries](https://github.com/apache/accumulo/blob/72110813107f14949bc082a109b4e35b06ea39ad/core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java#L1811-L1813) uses its fence range on the passed in extent.

### PR DESCRIPTION
**Pull request by cshannon**
This adds support for estimating overlapping entries in fenced RFiles. This works by clipping the KeyExtent provided with the fenced file range to find the overlapping section to use for computing the estimated entries.

A new clip() method was added to KeyExtent to make this easier to generate the new KeyExtent. Also, the range required must be the same type that KeyExtent uses when calling toDataRange()  (a more narrow row range) so that we can compute the overlapping section correctly. Fenced files allow other ranges but with our current use case fenced files will only be derived from KeyExtents so this should be fine. (Such as when doing merge operations)

I also added some tests, including missing RFile tests for estimating overlapping entries on non-fenced files.

This closes #4611

### CODE REVIEW

File: core/src/test/java/org/apache/accumulo/core/data/KeyExtentTest.java [hunk 1]
```diff
@@ -181,4 +182,42 @@ private KeyExtent writeAndReadFields(KeyExtent in) throws IOException {
...
+    clipped = ke.clip(range, true);
+    assertNull(clipped);
+    final Range r = range;
+    assertThrows(IllegalArgumentException.class, () -> ke.clip(r));
+
+    // Invalid row ranges
+    assertThrows(IllegalArgumentException.class,
+        () -> ke.clip(new Range("r_0002", true, "r_0004", true)));
+    assertThrows(IllegalArgumentException.class, () -> ke
+        .clip(new Range(new Key("r_0002", "cf_0002"), false, new Key("r_0003", "cf_0003"), true)));
```
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/data/KeyExtentTest.java:220]: Could add an additional test where the start row is empty string, this would exercise the code in isExclusiveKey that checks the key length.

File: core/src/main/java/org/apache/accumulo/core/util/RangeUtil.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ */
+package org.apache.accumulo.core.util;
+
+import org.apache.accumulo.core.data.ByteSequence;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Range;
+
+import com.google.common.base.Preconditions;
+
+public class RangeUtil {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/util/RangeUtil.java:27]: Could use the following name

```suggestion
public class RowRangeUtil {
```

File: core/src/main/java/org/apache/accumulo/core/dataImpl/KeyExtent.java [hunk 3]
```diff
@@ -388,6 +393,70 @@ public Range toMetaRange() {
...
+   * @see KeyExtent#clip(Range)
+   **/
+  public KeyExtent clip(Range range, boolean returnNullIfDisjoint) {
+    // A standard row range only requires an exclusive key but
+    // KeyExtent has a narrower definition of a row range and toDataRange()
+    // will always create exclusive start keys as well,
+    // so we need to restrict the range further
+    requireRowRange(range);
+    if (!range.isInfiniteStartKey()) {
+      Preconditions.checkArgument(isExclusiveKey(range.getStartKey()),
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/dataImpl/KeyExtent.java:439]: If want requireRowRange to be really generic then should also move the isExclusiveKey check for the end key outside to this function.  Looking at the range constructors for rows, it does not always append a 0x00 to the row key for all row ranges.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/RFileTest.java [hunk 4]
```diff
@@ -1512,6 +1512,39 @@ public void testReseekUnconsumed() throws Exception {
...
 
+  @Test
+  public void testEstimateOverlappingEntries() throws IOException {
+    TestRFile trf = new TestRFile(conf);
+    // lower block sizes so estimates are closer
+    trf.openWriter(true, 100, 100);
+
+    // generate 1024 entries
+    int count = 0;
+    for (int row = 0; row < 4; row++) {
```
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/RFileTest.java:1523]: Could this code to write data be pushed up into AbstractRfileTest so it could be shared w/ FencedRfileTest?

File: core/src/main/java/org/apache/accumulo/core/dataImpl/KeyExtent.java (hunk 3 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/dataImpl/KeyExtent.java:439]: Well, the original intent of requireRowRange was for no-chop merge as there were issues when not requiring that. I don't remember off the top of my head but it was added for a reason to require the end key to be exclusive when we store entries in metadata. So if we remove the exclusive key check then we'd need to add something to still enforce that for TabletFile and StoredTabletFile i think.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/RFileTest.java (hunk 4 above)
Comment (cshannon) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/RFileTest.java:1523]: We could, the only reason why I didn't was that the fencing test already has something to do a similar thing but I could likely refactor it to have it shared in the Abstract test (or at least try to)

Comment (keith-turner) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/RFileTest.java:1523]: Please ignore the comment if it does not make sense.  I was not sure about the class hierarchy w/ the test and code sharing.

Comment (cshannon) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/RFileTest.java:1523]: I think i'll leave things how they are for now since the fenced test uses another method already to write the test data

**Review by keith-turner — APPROVED** [review]

**Comment by cshannon** [comment]
requireRowRange has been renamed to `requireKeyExtentData` range and the checks have been more narrowed. Everywhere we are using fencing (currently) the ranges are derived from KeyExtent so it makes sense to validate. The actual RFiles can still use ranges that are not KeyExtent data ranges but `StoredTabletFile` will verify the correct format and we require that format when computing estimated overlapping entries

**Comment by cshannon** [comment]
The sunny ITs all passed for main and also when merging forward into elasticity. I will keep an eye on the full ITs and fix anything else that might have broken as there were some validation changed in this PR that could require some more test fixes.

**Comment by cshannon** [comment]
> The sunny ITs all passed for main and also when merging forward into elasticity. I will keep an eye on the full ITs and fix anything else that might have broken as there were some validation changed in this PR that could require some more test fixes.

The full IT build had 7 test failures but all are quick fixes. Fixed in 73125e9
